File: windows-app/AIClockBridge/ModelPricing.cs
using System.Text.Json;

namespace AIClockBridge;

readonly record struct TokenPrice(
    double Input, double CachedInput, double CacheWrite, double CacheWrite1h, double Output,
    double? PriorityInput = null, double? PriorityCachedInput = null, double? PriorityOutput = null);

/// Offline-first model price catalog. Prices are USD/token. The bundled table
/// works immediately; LiteLLM's public catalog refreshes it in the background.
sealed class ModelPricing
{
    public static readonly ModelPricing Shared = new();

    static readonly Uri Source = new(
        "https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json");
    static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(12) };
    readonly object _lock = new();
    readonly string _cachePath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "AIClockBridge", "model-prices.json");
    readonly Dictionary<string, TokenPrice> _prices = new(StringComparer.OrdinalIgnoreCase);

    ModelPricing()
    {
        foreach (var pair in Fallback) _prices[pair.Key] = pair.Value;
        try { Merge(File.ReadAllBytes(_cachePath)); } catch { }
    }

    public void Refresh() => _ = Task.Run(async () =>
    {
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(12));
            using var response = await Http.GetAsync(Source, HttpCompletionOption.ResponseHeadersRead, cts.Token);
            response.EnsureSuccessStatusCode();
            var data = await ReadLimited(response.Content, 16 * 1024 * 1024, cts.Token);
            Merge(data);
            Directory.CreateDirectory(Path.GetDirectoryName(_cachePath)!);
            await File.WriteAllBytesAsync(_cachePath, data);
        }
        catch { }
    });

    static async Task<byte[]> ReadLimited(HttpContent content, int limit, CancellationToken token)
    {
        if (content.Headers.ContentLength > limit) throw new InvalidDataException("model price catalog too large");
        await using var stream = await content.ReadAsStreamAsync(token);
        using var output = new MemoryStream(Math.Min(limit, (int)(content.Headers.ContentLength ?? 4096)));
        var buffer = new byte[32 * 1024];
        while (true)
        {
            var read = await stream.ReadAsync(buffer, token);
            if (read == 0) break;
            if (output.Length + read > limit) throw new InvalidDataException("model price catalog too large");
            output.Write(buffer, 0, read);
        }
        return output.ToArray();
    }

    public TokenPrice? Price(string rawModel, bool priority)
    {
        var model = (rawModel ?? "").ToLowerInvariant();
        var slash = model.LastIndexOf('/');
        if (slash >= 0) model = model[(slash + 1)..];
        var normalized = model;
        foreach (var suffix in new[] { "-extra-high", "-xhigh", "-ultra", "-medium", "-high", "-none", "-low", "-max" })
        {
            if (!model.EndsWith(suffix)) continue;
            normalized = model[..^suffix.Length];
            break;
        }

        TokenPrice? result;
        lock (_lock)
        {
            if (_prices.TryGetValue(model, out var exact)) result = exact;
            else if (_prices.TryGetValue(normalized, out var basePrice)) result = basePrice;
            else result = FamilyPrice(normalized);
        }
        if (!priority || result == null) return result;

        var p = result.Value;
        double? multiplier = p.PriorityInput.HasValue || p.PriorityOutput.HasValue ? 1
            : model.StartsWith("gpt-5.5") || model.StartsWith("gpt-5.6") ? 2.5
            : model.StartsWith("gpt-5") ? 2 : null;
        if (!multiplier.HasValue) return null;
        return new TokenPrice(
            p.PriorityInput ?? p.Input * multiplier.Value,
            p.PriorityCachedInput ?? p.CachedInput * multiplier.Value,
            p.CacheWrite * multiplier.Value, p.CacheWrite1h * multiplier.Value,
            p.PriorityOutput ?? p.Output * multiplier.Value,
            p.PriorityInput, p.PriorityCachedInput, p.PriorityOutput);
    }

    TokenPrice? FamilyPrice(string model)
    {
        string key = model.StartsWith("claude-opus-4-") ? "claude-opus-4"
            : model.StartsWith("claude-sonnet-4-") ? "claude-sonnet-4"
            : model.StartsWith("claude-haiku-4-") ? "claude-haiku-4"
            : model.StartsWith("gpt-5-codex-") ? "gpt-5-codex" : null;
        return key != null && _prices.TryGetValue(key, out var price) ? price : null;
    }

    void Merge(byte[] data)
    {
        using var doc = JsonDocument.Parse(data);
        var parsed = new Dictionary<string, TokenPrice>(StringComparer.OrdinalIgnoreCase);
        foreach (var model in doc.RootElement.EnumerateObject())
        {
            var v = model.Value;
            if (!Number(v, "input_cost_per_token", out var input)
                || !Number(v, "output_cost_per_token", out var output)) continue;
            parsed[model.Name] = new TokenPrice(
                input, Number(v, "cache_read_input_token_cost") ?? input,
                Number(v, "cache_creation_input_token_cost") ?? input,
                Number(v, "cache_creation_input_token_cost_above_1hr") ?? input * 2,
                output, Number(v, "input_cost_per_token_priority"),
                Number(v, "cache_read_input_token_cost_priority"),
                Number(v, "output_cost_per_token_priority"));
        }
        if (parsed.Count == 0) return;
        lock (_lock) foreach (var pair in parsed) _prices[pair.Key] = pair.Value;
    }

    static bool Number(JsonElement obj, string name, out double value)
    {
        value = 0;
        return obj.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.Number
            && v.TryGetDouble(out value) && double.IsFinite(value) && value >= 0 && value <= 1;
    }

    static double? Number(JsonElement obj, string name)
        => Number(obj, name, out var value) ? value : null;

    static readonly Dictionary<string, TokenPrice> Fallback = new(StringComparer.OrdinalIgnoreCase)
    {
        ["gpt-5"] = new(0.00000125, 0.000000125, 0.00000125, 0.0000025, 0.00001,
            0.0000025, 0.00000025, 0.00002),
        ["gpt-5-codex"] = new(0.00000125, 0.000000125, 0.00000125, 0.0000025, 0.00001),
        ["gpt-5.6-sol"] = new(0.000005, 0.0000005, 0.000005, 0.00001, 0.00003,
            0.00001, 0.000001, 0.00006),
        ["claude-opus-4"] = new(0.000015, 0.0000015, 0.00001875, 0.00003, 0.000075),
        ["claude-sonnet-4"] = new(0.000003, 0.0000003, 0.00000375, 0.000006, 0.000015),
        ["claude-haiku-4"] = new(0.000001, 0.0000001, 0.00000125, 0.000002, 0.000005),
    };
}

File: mac-app/Sources/AIClockBridge/ModelPricing.swift
import Foundation

struct TokenPrice {
    let input: Double
    let cachedInput: Double
    let cacheWrite: Double
    let cacheWrite1h: Double
    let output: Double
    let priorityInput: Double?
    let priorityCachedInput: Double?
    let priorityOutput: Double?
}

/// Small, dependency-free pricing catalog. A bundled fallback keeps totals
/// useful offline; LiteLLM's public catalog refreshes exact model prices at
/// launch and is cached locally for the next offline run.
final class ModelPricing {
    static let shared = ModelPricing()

    private let lock = NSLock()
    private var prices: [String: TokenPrice] = [:]
    private let source = URL(string: "https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json")!
    private let cacheURL: URL

    private init() {
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("AIClockBridge", isDirectory: true)
        try? FileManager.default.createDirectory(at: support, withIntermediateDirectories: true)
        cacheURL = support.appendingPathComponent("model-prices.json")
        prices = Self.fallback
        if let data = try? Data(contentsOf: cacheURL) { merge(data) }
    }

    func refresh() {
        var request = URLRequest(url: source)
        request.timeoutInterval = 12
        URLSession.shared.dataTask(with: request) { [weak self] data, response, _ in
            guard let self, let data,
                  (response as? HTTPURLResponse)?.statusCode == 200 else { return }
            self.merge(data)
            try? data.write(to: self.cacheURL, options: .atomic)
        }.resume()
    }

    func price(for rawModel: String, priority: Bool) -> TokenPrice? {
        var model = rawModel.lowercased()
        if let slash = model.lastIndex(of: "/") { model = String(model[model.index(after: slash)...]) }
        let effortSuffixes = ["-extra-high", "-xhigh", "-ultra", "-medium", "-high", "-none", "-low", "-max"]
        let normalized = effortSuffixes.first(where: { model.hasSuffix($0) })
            .map { String(model.dropLast($0.count)) } ?? model
        lock.lock()
        defer { lock.unlock() }
        let base = prices[model] ?? prices[normalized] ?? fallbackFamilyPrice(normalized)
        guard let p = base, priority else { return base }
        let multiplier: Double?
        if p.priorityInput != nil || p.priorityOutput != nil { multiplier = 1 }
        else if model.hasPrefix("gpt-5.5") || model.hasPrefix("gpt-5.6") { multiplier = 2.5 }
        else if model.hasPrefix("gpt-5") { multiplier = 2 }
        else { multiplier = nil }
        guard let multiplier else { return nil }
        return TokenPrice(input: p.priorityInput ?? p.input * multiplier,
                          cachedInput: p.priorityCachedInput ?? p.cachedInput * multiplier,
                          cacheWrite: p.cacheWrite * multiplier,
                          cacheWrite1h: p.cacheWrite1h * multiplier,
                          output: p.priorityOutput ?? p.output * multiplier,
                          priorityInput: p.priorityInput,
                          priorityCachedInput: p.priorityCachedInput,
                          priorityOutput: p.priorityOutput)
    }

    private func fallbackFamilyPrice(_ model: String) -> TokenPrice? {
        if model.hasPrefix("claude-opus-4-") { return prices["claude-opus-4"] }
        if model.hasPrefix("claude-sonnet-4-") { return prices["claude-sonnet-4"] }
        if model.hasPrefix("claude-haiku-4-") { return prices["claude-haiku-4"] }
        if model.hasPrefix("gpt-5-codex-") { return prices["gpt-5-codex"] }
        return nil
    }

    private func merge(_ data: Data) {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
        var parsed: [String: TokenPrice] = [:]
        for (name, value) in root {
            guard let v = value as? [String: Any],
                  let input = (v["input_cost_per_token"] as? NSNumber)?.doubleValue,
                  let output = (v["output_cost_per_token"] as? NSNumber)?.doubleValue else { continue }
            parsed[name.lowercased()] = TokenPrice(
                input: input,
                cachedInput: (v["cache_read_input_token_cost"] as? NSNumber)?.doubleValue ?? input,
                cacheWrite: (v["cache_creation_input_token_cost"] as? NSNumber)?.doubleValue ?? input,
                cacheWrite1h: (v["cache_creation_input_token_cost_above_1hr"] as? NSNumber)?.doubleValue
                    ?? input * 2,
                output: output,
                priorityInput: (v["input_cost_per_token_priority"] as? NSNumber)?.doubleValue,
                priorityCachedInput: (v["cache_read_input_token_cost_priority"] as? NSNumber)?.doubleValue,
                priorityOutput: (v["output_cost_per_token_priority"] as? NSNumber)?.doubleValue
            )
        }
        guard !parsed.isEmpty else { return }
        lock.lock()
        prices.merge(parsed) { _, new in new }
        lock.unlock()
    }

    private static let fallback: [String: TokenPrice] = [
        "gpt-5": .init(input: 0.00000125, cachedInput: 0.000000125, cacheWrite: 0.00000125,
                       cacheWrite1h: 0.0000025,
                       output: 0.00001, priorityInput: 0.0000025,
                       priorityCachedInput: 0.00000025, priorityOutput: 0.00002),
        "gpt-5-codex": .init(input: 0.00000125, cachedInput: 0.000000125, cacheWrite: 0.00000125,
                             cacheWrite1h: 0.0000025,
                             output: 0.00001, priorityInput: nil,
                             priorityCachedInput: nil, priorityOutput: nil),
        "gpt-5.6-sol": .init(input: 0.000005, cachedInput: 0.0000005, cacheWrite: 0.000005,
                             cacheWrite1h: 0.00001,
                             output: 0.00003, priorityInput: 0.00001,
                             priorityCachedInput: 0.000001, priorityOutput: 0.00006),
        "claude-opus-4": .init(input: 0.000015, cachedInput: 0.0000015, cacheWrite: 0.00001875,
                               cacheWrite1h: 0.00003,
                               output: 0.000075, priorityInput: nil,
                               priorityCachedInput: nil, priorityOutput: nil),
        "claude-sonnet-4": .init(input: 0.000003, cachedInput: 0.0000003, cacheWrite: 0.00000375,
                                 cacheWrite1h: 0.000006,
                                 output: 0.000015, priorityInput: nil,
                                 priorityCachedInput: nil, priorityOutput: nil),
        "claude-haiku-4": .init(input: 0.000001, cachedInput: 0.0000001, cacheWrite: 0.00000125,
                                cacheWrite1h: 0.000002,
                                output: 0.000005, priorityInput: nil,
                                priorityCachedInput: nil, priorityOutput: nil),
    ]
}
